New Object domain

LinerSu opened the pull request:
Introduce a new object domain based on DSA node and cache operation.

LinerSu committed:
feat(domain): add a class to wrap domain used in ODI map

LinerSu committed:
feat(domain): introduce new object domain based on dsa node

LinerSu committed:
feat(test): add object domain implementation type

LinerSu committed:
test(object-dom): adapt unit tests for new object domain

agurfinkel commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,1303 @@
+    // The following map update is not necessary because m_flds_rep_map
+    // is fixed after intrinsic calls.
+    // Here, in case to keep operation sound.
+    flds_rep_map_t out_flds_rep_map;
```
this is wrong. `flds_rep_map` must be a shared map between all values, not a copy!

agurfinkel commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,1303 @@
+    odi_map_t out_odi_map(is_join ? left.m_odi_map | right.m_odi_map
+                                  : left.m_odi_map || right.m_odi_map);
+
+    // The following map update is not necessary because m_flds_rep_map
```
wrong, see comment above.

agurfinkel commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,1303 @@
+  void get_rgn_vec(variable_t rep, variable_vector_t &rgn_vec) {
+    for (auto kv : m_flds_rep_map) {
+      if (kv.second == rep) {
+        rgn_vec.push_back(kv.second);
```
`kv.first`? 

Currently, the function fills `rgn_vec` with `rep`

agurfinkel commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,1303 @@
+    }
+  }
+
+  variable_t get_object_rep_by_rgn(variable_t rgn) {
```
rename, `get_oid`

agurfinkel commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,1303 @@
+  }
+
+  template <class DOM>
+  void assign_interval(DOM &dom, const variable_t v, interval_t ival) {
```
this is a helper function. It should not be part of the class.

agurfinkel commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,1303 @@
+    variable_vector_t rgn_vec;
+    base_abstract_domain_t tmp_base(m_base_dom);
+    tmp_base.project(rgn_vec);
+    // Be careful of the following operation if base dom and object dom are
```
you can use a `static_assert` to check for what the comment is saying, or implement a cast operation

agurfinkel commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,1303 @@
+  void move_singleton_to_odi_map(variable_t rep) {
+    variable_vector_t rgn_vec;
+    base_abstract_domain_t tmp_base(m_base_dom);
+    tmp_base.project(rgn_vec);
```
`rgn_vec` is empty at this point. Probably forgot to initialize it.

agurfinkel commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,1303 @@
+    error_if_not_rgn(x);
+
+    if (!is_bottom()) {
+      auto it = m_flds_rep_map.find(x);
```
this needs to be cleaned up. there are functions for this functionality

agurfinkel commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,1303 @@
+  }
+
+  // add constraint into object domain
+  void add_cons_for_objects(const linear_constraint_t &cst) {
```
this function is a mess. better remove it before we have a good use for it.

agurfinkel commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,1303 @@
+    auto obj_info = m_obj_info_env[rep];
+
+    const small_range &num_refs = obj_info.count_dom();
+    if (num_refs.is_zero()) {
```
how is this possible? Reading from an object with 0 instances?

agurfinkel commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,1303 @@
+      m_base_dom -= res;
+    } else if (num_refs.is_one()) { // singleton object
+      // strong read
+      m_base_dom.assign(res, rgn);
```
this should crash. `rgn` is not a variable known to `m_base_dom`. In the worst case, `rgn` is mapped to an index that is known to `m_base_dom` but represents a different variable. Then this will not crash but will produce very strange results.

agurfinkel commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,1303 @@
+    } else {
+      // read from odi map
+      // weak read
+      auto obj_dom = m_odi_map[rep];
```
this is better.

Here, and everywhere, you should be very careful with copy-assignments. I believe that 

```
auto obj_dom = ...
```
is very expensive and should be
```
auto &obj_dom = ...
```

However, you sometimes re-assign `obj_dom`, so you can't use a reference all the time, but should only copy when necessary.

agurfinkel commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,1303 @@
+
+    const small_range &num_refs = obj_info.count_dom();
+    if (num_refs.is_zero()) {
+      m_base_dom -= rgn;
```
I don't think this possible or meaningful

agurfinkel commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,1303 @@
+      if (val.is_constant())
+        m_base_dom.assign(rgn, val.get_constant());
+      else
+        m_base_dom.assign(rgn, val.get_variable());
```
this is wrong. see comment above

agurfinkel reviewed (changes requested):
see comments

agurfinkel commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,1286 @@
+    auto obj_dom_ptr = m_odi_map.find(rep);
+    field_abstract_domain_t *res_obj_dom_ptr;
+    if (obj_dom_ptr) { // found in odi map, create new one by copying old one
+      res_obj_dom_ptr = new field_abstract_domain_t((*obj_dom_ptr));
```
the rest of the code is ok, just rewrite the code so that the allocation is on the stack

agurfinkel reviewed (changes requested):
some memory related issues. the rest is good. 

does this run on our benchmarks? How does it compare with vstte?

caballa commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,1286 @@
+      m_base_dom -= res; // forget res means res == top
+      // the requirment 3.1 is satisfied
+    } else {
+      field_abstract_domain_t *obj_dom_ref_no_const =
```
There are two issues with returning an interval. First, the underlying domain might not keep any interval representation. That's why the returned interval is by value. Second,  the method `interval_t operator[](...)` is not marked as `const` because some domains (e.g., octagons) needs to apply normalization before producing the interval. I could have a second `const` version that cannot normalize.

agurfinkel replied:
A C++ convention is to provide a method called `at()` that is same as `operator[]` but const.

The domain should not expose its internal normalization through `const`. It should hide it using `mutable`. 
At the very least, you want to provide a const `at` method, that, for example, does `const cast` under the hood.

LinerSu commented:
@agurfinkel About how object domain compared with VSTTE, I walked through this example.
```
entry:
  region_init(V_i:region(int));
  region_init(V_x:region(int));
  region_init(V_y:region(int));
  i := make_ref(V_i:region(int),4,as_0);
  x := make_ref(V_x:region(int),4,as_1);
  y := make_ref(V_y:region(int),4,as_2);
  store_to_ref(V_i:region(int),i:ref,0:int32);
  store_to_ref(V_x:region(int),x:ref,1:int32);
  store_to_ref(V_y:region(int),y:ref,0:int32);
  goto bb1;
bb1:
  goto bb1_t,bb1_f;
bb1_t:
  *i:int32 := load_from_ref(V_i:region(int),i:ref);
  assume(*i <= 99);
  goto bb2;
bb2:
  *x:int32 := load_from_ref(V_x:region(int),x:ref);
  *y:int32 := load_from_ref(V_y:region(int),y:ref);
  *x = *x+*y;
  store_to_ref(V_x:region(int),x:ref,*x:int32);
  *y:int32 := load_from_ref(V_y:region(int),y:ref);
  *y = *y+1;
  store_to_ref(V_y:region(int),y:ref,*y:int32);
  *i:int32 := load_from_ref(V_i:region(int),i:ref);
  *i = *i+1;
  store_to_ref(V_i:region(int),i:ref,*i:int32);
  goto bb2_end;
bb2_end:
  goto bb1;
bb1_f:
  *i:int32 := load_from_ref(V_i:region(int),i:ref);
  assume(-*i <= -100);
  goto bb3;
bb3:
  *x:int32 := load_from_ref(V_x:region(int),x:ref);
  *y:int32 := load_from_ref(V_y:region(int),y:ref);
  goto ret;
ret:
  assert(-*x+*y <= 0);
```
I use splitDBM for region domain and object domain, the VSTTE can prove assertion in `ret` block. The object domain does not.
The invariants for object domain are:
```
entry={}
bb1=Base = {i-objA<=0, x-objA<=4, y-objA<=8, objA-i<=0, x-i<=4, y-i<=8, objA-x<=-4, i-x<=-4, y-x<=4, x-y<=-4, objA-y<=-8, i-y<=-8},
Object( V_i V_x V_y ) = {V_i -> [0, +oo], V_x -> [1, +oo], V_y -> [0, +oo]}

bb1_f=Base = {i-objA<=0, x-objA<=4, y-objA<=8, objA-i<=0, x-i<=4, y-i<=8, objA-x<=-4, i-x<=-4, y-x<=4, x-y<=-4, objA-y<=-8, i-y<=-8},
Object( V_i V_x V_y ) = {V_i -> [0, +oo], V_x -> [1, +oo], V_y -> [0, +oo]}

bb3=Base = {*i -> [100, +oo], i-objA<=0, x-objA<=4, y-objA<=8, objA-i<=0, x-i<=4, y-i<=8, objA-x<=-4, i-x<=-4, y-x<=4, x-y<=-4, objA-y<=-8, i-y<=-8},
Object( V_i V_x V_y ) = {V_i -> [0, +oo], V_x -> [1, +oo], V_y -> [0, +oo]}

ret=Base = {*i -> [100, +oo], *x -> [1, +oo], *y -> [0, +oo], i-objA<=0, x-objA<=4, y-objA<=8, objA-i<=0, x-i<=4, y-i<=8, objA-x<=-4, i-x<=-4, y-x<=4, x-y<=-4, objA-y<=-8, i-y<=-8},
Object( V_i V_x V_y ) = {V_i -> [0, +oo], V_x -> [1, +oo], V_y -> [0, +oo]}

bb1_t=Base = {i-objA<=0, x-objA<=4, y-objA<=8, objA-i<=0, x-i<=4, y-i<=8, objA-x<=-4, i-x<=-4, y-x<=4, x-y<=-4, objA-y<=-8, i-y<=-8},
Object( V_i V_x V_y ) = {V_i -> [0, +oo], V_x -> [1, +oo], V_y -> [0, +oo]}

bb2=Base = {*i -> [0, 99], i-objA<=0, x-objA<=4, y-objA<=8, objA-i<=0, x-i<=4, y-i<=8, objA-x<=-4, i-x<=-4, y-x<=4, x-y<=-4, objA-y<=-8, i-y<=-8},
Object( V_i V_x V_y ) = {V_i -> [0, +oo], V_x -> [1, +oo], V_y -> [0, +oo]}
```
Here, we lost `V_i == V_x` in fields domain or `*i == *x` in base domain. One reason I think is we lost `V_i == *i` and `V_x == *x` because the `load_ref` or `store_ref` is assigning intervals instead of keeping equality in a domain.

agurfinkel commented:
@LinerSu looks cool. At this point, I only care about performance. Precision is expected to be very low.

LinerSu committed:
fix(domain): fix load/store ref implementation

LinerSu committed:
feat(domain): treat no-op if region does not belong to an abastract object

LinerSu committed:
feat(domain): add implementation for extraction of properties

LinerSu committed:
feat(object-dom): move singleton objects into base domain

caballa commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,1723 @@
+          right.m_obj_info_env.at(rep).refcount_val();
+      if (left_num_refs.is_one() &&
+          right_num_refs == small_range::oneOrMore()) {
+        join_singleton_with_non_singleton(rep, left.m_base_dom, out_odi_map);
```
you need to apply widening if `is_join` is false

caballa commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,1723 @@
+        join_singleton_with_non_singleton(rep, left.m_base_dom, out_odi_map);
+      } else if (right_num_refs.is_one() &&
+                 left_num_refs == small_range::oneOrMore()) {
+        join_singleton_with_non_singleton(rep, right.m_base_dom, out_odi_map);
```
same here

caballa commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,1723 @@
+      if (left_num_refs.is_one() &&
+          right_num_refs == small_range::oneOrMore()) {
+        meet_non_singleton_with_singleton(rep, out_base_dom, right.m_odi_map);
+        field_abstract_domain_t out_obj_dom; // top
```
Here it seems that we are losing a lot of precision by assigning top. why? 
meet should be able to refine things and improve precision.

caballa commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,1723 @@
+      // region belongs to an abstract object
+
+      // retrieve an abstract object info
+      auto obj_info = m_obj_info_env.at(*id_opt);
```
As a general comment, note that `separate_domain` already provides `const Value* find(const Key &k)`. So in cases like this you can avoid copies

caballa commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,1723 @@
+    tmp_base.project(flds_vec);
+
+    // Be careful of the following operation if type of base dom and object dom
+    // are different
```
This is something we can check with `static_assert` at the beginning of the class, right?

LinerSu replied:
I used to have it in a function:
https://github.com/LinerSu/crab/blob/5018d40d7b7cca55ad05225e0d2af848f3d42125/include/crab/domains/object_domain.hpp#L547-L552
Now I see the `static_assert` can be a member declaration. I will change that.

LinerSu committed:
fix(object-domain): apply widen/narrow for singleton reduction

LinerSu committed:
refactor(object-domain): rearrange APIs based on the type of ops

igcontreras commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,1811 @@
+    bool res = left.m_base_dom <= right.m_base_dom;
+    CRAB_LOG("object", crab::outs() << "Result4=" << res << "\n";);
+
+    res = left.m_odi_map <= right.m_odi_map;
```
this should be `res = res && (left.m_odi_map <= right.m_odi_map);`

LinerSu committed:
fix(obj-dom): fix join/widen op on singleton to non-singleton

LinerSu committed:
fix(obj-dom): fix leq op

LinerSu committed:
feat(object-domain): handle object not introduced by intrinsic calls

LinerSu committed:
feat(object-domain): add forget implementation

LinerSu committed:
fix(obj-dom): fix update object info map

In crab IR, region initialization / copy occurs before intrinsic calls.
The object info is required for each abstract object so that
we need to initialize that info for object not informed by
intrinsic calls.

LinerSu committed:
lib(term): add a term operator value generator

LinerSu committed:
feat(object): add a object info class to keep object's info

LinerSu committed:
feat(object): add implementation of cache, address, and regs doms

The abstract domain is based on memory abstraction which models
memory architecture with cache. Each abstract object contains a
cache that precisely materializes memory load / store. The cache
only records the most recently used by load/store instructions.
The current cache size is one so that any loads / stores for
non mru object require to commit and update cache subdomain.

In the implementation, we also includes an alias analysis that abstracts
references alias properties. That is, we only know two references are
alias according to the DSA analysis. Such info are stored in an EUF domain.

LinerSu committed:
test(object-dom): adapt unit tests for new object domain

LinerSu committed:
fix(object): fix join/meet odi map if odi is top

LinerSu committed:
refactor(object): adpat new crab APIs

LinerSu committed:
feat(object): assign address symbol for int to ref

LinerSu committed:
test(object): add test for join op

igcontreras commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,2378 @@
+      const object_info_t &l_obj_info = kv.second;
+      const small_range &l_num_refs = l_obj_info.refcount_val();
+      auto r_obj_info_ref = right.m_obj_info_env.find(id);
+      if (!r_obj_info_ref) {
```
If an object is not in `right` does it represent that for that object we have top? should the object be forgotten in `this`?

igcontreras commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,2378 @@
+  // Perform *this = join(*this, right)
+  void self_join(const object_domain_t &right) {
+
+    // The join operation is sound if:
```
what does 'if' here mean? Are the invariants assumptions of `this` and `right`?

igcontreras commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,2378 @@
+    variable_t mru_base = get_base_addr_or_fail(id);
+    const usymb_t *l_term_ptr = l_addrs_dom.get_term(mru_base);
+    const usymb_t *r_term_ptr = r_addrs_dom.get_term(mru_base);
+    if (l_term_ptr && r_term_ptr) {
```
Is this code to check if the 2 uninterpreted constants are equal? if `not(a < b)` and `not(b < a)` it could be that `a = b` but it could also be that they are not comparable.

igcontreras commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,2378 @@
+  }
+
+  bool test_two_addrs_equality(const variable_t &x, const variable_t &y) {
+    const usymb_t *x_term_ptr =
```
same question as in `mru_refer_same_object`. Part of this function could be reused in `mru_refer_same_object`

igcontreras commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,2378 @@
+    assert(obj_info_ref);
+    // Be careful of the following operation if type of base dom and object
+    // dom are different
+    if (prod_ref) {
```
This `if` can be moved to line 958

igcontreras commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,2378 @@
+
+    // Be careful of the following operation if type of base dom and object dom
+    // are different
+    if (prod_ref) {
```
This `if` can be moved to line 986

igcontreras commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,2378 @@
+          if (res.get_type().is_reference()) {
+            // res is a reference, assign a fresh symbol into address dom
+            // create a fresh symbol to represent its base address
+            m_addrs_dom.set(get_or_insert_base_addr(res),
```
Why new symbol? It should be the same symbol as the base address of what is currently loaded in the cache

Commit:
fix(object): convert term to term op for comparison

We use UF domain to represent base addresses.
The caches in given two states refer the same MRU object if the base addresses of caches refer the same symbol in the uf domain.
The symbol is actually a term operator in crab. The comparison is based
on the value of operator but not the term.

LinerSu committed:
feat(object): add reduction operator between cache and base subdomains

Introduce a new redution operator (transfer function) that reduce equalities
between registers and fields in the cache. The reduction can be either adding
constraints for registers into base domain or adding constraints for fields
into corresponding cache domain.

LinerSu committed:
feat(object): add intrinsic method to perform reduction

The object domain can handle an intrinsic reduction named:
crab_intrinsic(do_reduction,@V_1:region,V_2:ref,V_3:bool);
where V_1 is a region and V_2 is a reference. The V_3 indicates
the direction for reduction. The value of V_3 is true(false) will
perform reduction from cache to base (from base to cache).

LinerSu committed:
refactor(object): add object domain version of select_ref

LinerSu committed:
feat(object): perform reduction under each transfer function

LinerSu committed:
refactor(object): use new api to compare two term operators

LinerSu committed:
test(object): complete join operator testing

Commit:
fix(object): fix assign address map in join/meet and treat noop when object is top

LinerSu committed:
fix(object): break register-field equality when assigning that reg

LinerSu committed:
feat(object): add a printing log for verbose printing

LinerSu committed:
refactor(ghost): make project and forget methods to const

LinerSu committed:
feat(object): use ghost variables to express reference with extra information

Since object domain can express relations between fields in more expressive
way, we also want to deal with pointer analysis such as is-dereferenceable.
The requirement for this check is using ghost variables to convey pointer
information: offset and size.

In object domain, now each typed variable such as region or reference
are now having at least one ghost variables with fixed naming.
This is also guarentee the efficiency for creating and manipulating
such variables.

LinerSu committed:
fix(object): ignore rhs region in select_ref if it is unknown region

LinerSu committed:
fear(object): add a domain flag to enable/disable reduction

LinerSu committed:
fix(params): fix order of initializating object params

LinerSu committed:
feat(object): implement forget, project, rename domain operators

Based on subdomains we have right now, we need to distinguish three
types of variables: variables in base domain, reference variables,
and fields variables used in odi map.

Once we have this classification, the operators can be performed based
on each subdomain: base domain, address domain, uf register domain,
and (summary domain, cache domain, uf field domain) in odi map.

LinerSu committed:
test: add object domain operator test

LinerSu committed:
profile: include statistics for moving singleton to odi map

igcontreras commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,3312 @@
+    if (get_obj_id(rgn) == boost::none) {
+      // if a region does not belong to an object, treat it as an object
+      // treat region as a field, as well as an object id
+      update_fields_id_map(rgn, rgn);
```
In these 2 functions you are doing `(*m_flds_id_map).find(rgn)` twice. See if this can be improved

igcontreras commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,3312 @@
+    // ref_store(ref, rgn, val);
+    // Definition: assign abstract value in val to rgn
+    //  rgn: a region variable used in object domain
+    //  res: a register variable used in base domain
```
`val`

igcontreras commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,3312 @@
+        }
+        invalidate_cache_if_miss((*id_opt), ref, rgn, reg_symb,
+                                 reg_offset_size_symb);
+        prod_ref = m_odi_map.find((*id_opt));
```
duplicated? (see `const odi_domain_product_t *prod_ref = m_odi_map.find((*id_opt));`)

igcontreras commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,3312 @@
+
+    if (!is_bottom()) {
+      // REDUCTION: perform reduction
+      perform_reduction();
```
why here?

igcontreras commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,3312 @@
+
+    if (!is_bottom()) {
+      // REDUCTION: perform reduction
+      perform_reduction();
```
If reducing is a problem, add a flag about what has not been reduced.

For example, if we have several operations with registers
```
x++;
y = y + x;
z = 2 * y;
```

right now we reduce after every operation. you can have a flag per object + base
dom of whether they have been modified since the last reduction.

When you reach a reduction point, do not transfer from objects that have not been modified since the last reduction.

Also, if the operations only involve register arithmetic, there is nothing to reduce until there is a store using those registers.

igcontreras commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,3312 @@
+    }
+  }
+
+  void update_fields_id_map(const variable_t &rgn, const obj_id_t &id) {
```
fields -> flds to be consistent?

LinerSu replied:
This function will be called if we evaluate either DSA intrinsic or region int statement in crabIR.

igcontreras commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,3312 @@
+  }
+
+  base_dom_variable_or_constant_t
+  rename_variable_or_constant(const variable_or_constant_t &v) {
```
is this equivalent to what is being done in the region domain?

igcontreras commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,3312 @@
+  //  (1) perform reduction by transfering regs-flds' constriants between cache
+  //  domain and base domain (2) clear UF field domain (3) clear cache domain
+  void
+  commit_cache_if_dirty(base_abstract_domain_t &base_dom,
```
if the object is top, just reset the cache/do nothing currently there is a lot of renaming/euf domain reasoning that could be fruitless

igcontreras commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,3312 @@
+    uf_fld_dom.rename(old_ghost_vars, new_ghost_vars);
+  }
+
+  void forget_a_field(const variable_t &v) {
```
`forget_a_field` -> `forget_fld`?

LinerSu committed:
refactor(object): optimize join op checking cases

LinerSu committed:
refactor(object): add constraints for offset ghost

LinerSu committed:
feat(object): implement new meet abstract operation &=

LinerSu committed:
feat(reduce): perform reduction if needs in intrinsic function

LinerSu committed:
fix(object): fix join / meet for singleton objects

This commit fixes the issue of join / meet when two states
contain singleton objects. Two states with singleton objects
where they might refer the same object (with same address),
or they are different. Therefore, join of two singleton objects
will become a non-singleton object if they are different.

LinerSu committed:
refactor(object): remove unused code

LinerSu committed:
fix: update object info after join / meet

If odi map is used, the object is not singleton. Therefore,
any join / meet over non-singleton objects require to commit caches
if necessary. If committed, we also need to reset caches as empty.

LinerSu committed:
feat(object-domain): flag to move singleton to base domain

- change global object-id map (locate) function to local
- add efficient join / meet operations on odi map

LinerSu committed:
fix(reduction): add essential reduction

caballa committed:
feat: add new abstract operations for calls [ci skip]

LinerSu committed:
feat(inter-analysis): integreate dsa info into callsite information

Giving a method to group region variables passed/returned between
caller and callee based on dsa intrinsics. The information is precomputed
before the analysis and stored into callsite_info class for further
abstract operations of calls.

LinerSu committed:
feat(inter): use call graph nodes iterator to retrieve all cfgs

gretadolcetti committed:
chore: updated download link for PPLite 0.9

gretadolcetti committed:
feat(domains): added implementation of decoupled domains

The decoupled domain allows crab to use two different domains during
the two different phases of the analysis.  In the descending
(narrowing) phase it is possible to use an abstract domain which is
more precise than the domain used in the ascending (widening) phase.

The decoupled domains require two new abstract operations
is_asc_phase() and set_phase() in order to manage ascending and
descending phases. By default, start in *descending* (i.e., more
precise) phase. The interleaved fixpoint iterator has been also
modified to notify the underlying domains when there is change of
(ascending or descending) phase.

Implemented by Greta Dolcetti and Enea Zaffanella.

LinerSu committed:
feat(domain): a symbolic variable equality domain

A symbolic equality domain is an abstract domain which represents equalities
used in analyses for allocation-site abstraction or domain reduction.
In short, giving two program variables are known to hold equal values in concrete semantics
if the analyzer determines that the variables hold equal symbolic term.

LinerSu committed:
feat(domain): an environement map domain storing object odi

A special seperate domain for odi map. Especially, it covers
different cases during domain operations such as join or meet.

LinerSu committed:
refactor(domain): update domain to use odi_map domain and symbolic var eq domain

LinerSu committed:
fix(domain): add missed condition checks on add(x, y)

LinerSu committed:
profile: include statistics for odi map domain

LinerSu committed:
fix(object): add a flag to disable all reduction

igcontreras commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,3681 @@
+  }
+
+  /***************** ODI map operations *****************/
+  /// @brief move singleton object into odi map
```
update comment

igcontreras commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,3681 @@
+  }
+
+  /***************** Cache operations *****************/
+  // commit cache contents into summary
```
if the object is singleton, you can do a reassign instead of join. I don't know if this case actually happens, but if executing the flush cache intrinsic causes the cache to be flushed and then the cache is updated you would lose precision here

igcontreras commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,3681 @@
+    get_obj_flds_with_ghost(id, obj_flds);
+    // (1) perform reduction by transfering regs-flds' constriants between cache
+    //  domain and base domain
+    reduce_regs_flds_between_cache_and_base(base_dom, eq_regs_dom, prod,
```
is this required for soundness?

LinerSu replied:
No, I added this because we no longer keep equality between fields and registers once the cache is flushed. If we commented this out. We just lost numerical properties for equality (i.e. imprecise but still sound).

igcontreras commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,3681 @@
+    eq_fld.set_to_top();
+  }
+
+  /// @brief
```
doc

igcontreras commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,3681 @@
+      return less_than_eq(left, right);
+    };
+
+    if (commit_is_required() && o.commit_is_required()) {
```
see my comment for `commit_is_required`

igcontreras reviewed (changes requested):
I looked at `object_domain.hpp` I did not finished. I think some of the performance comments will be addressed when we have the reduction flags that we talked about.

I see some comments from my last review are still here. Please either reply to them or resolve them so we can track progress better.

caballa committed:
fix(test): enable inter operations for apron/elina

caballa committed:
perf(varname): avoid calls to + and copies with std::string

caballa committed:
chore: added comments

caballa committed:
feat(checker): new get_all_checks() method

caballa committed:
perf(domains): use of stats

We minimize string allocation and copies.  Also, we add some macros
that allow each domain to turn on or off statistics collection which
adds a counter and a timer per operation. If stats collection is off
then no string allocation or string copy should take place.

For instance, the region domain enables stats by adding

#define REGION_DOMAIN_SCOPED_STATS(NAME) CRAB_DOMAIN_SCOPED_STATS(NAME, 1)

And the interval domain disables stats by adding

#define INTERVALS_DOMAIN_SCOPED_STATS(NAME) CRAB_DOMAIN_SCOPED_STATS(NAME, 0)

The macro CRAB_DOMAIN_SCOPED_STATS is defined in
crab/domains/abstract_domain_macros.def

In spite of all optimizations, the most efficient thing is to disable
stats collection. By default, only the fixpoint solver and the region
domain turn on stats. This is okay because stats gathering is a
feature intended only for developers. If needed then the developer can
turn on more stats timers and counters and recompile the code.

caballa committed:
chore: remove use of std::strlen

LinerSu commented on include/crab/domains/object_domain.hpp:
```diff
@@ -0,0 +1,3681 @@
+    }
+    // NOTE: copy the following value is required because
+    // updating the odi map is construcing a new tree based on previous tree
+    odi_domain_product_t res_prod = *prod_ref;
```
Yes, even if a singleton object is in the base because the ODI map stores a pair of `<object id, <info, value>>` where the info stores cache status and reference counting. If a singleton in base, the `value` is a top. Copying a top value is cheap.

LinerSu committed:
refactor(params): change doamin reduction params with multiple options

object.reduction_level=FULL_REDUCTION to perform reduction before
each transfer function;
object.reduction_level=REDUCTION_BEFORE_CHECK to perform
reduction only before assertion;
object.reduction_level=NO_REDUCTION for no reduction.

LinerSu committed:
refactor(domain): refactor domains for inferring object invariants

- `object_domain`: avoid deep copy during updating odi values;
support new statistics; refactor cache operations.
- `odi_map_domain`: perform cache flushing before join / meet
each odi value; refactor cache operations.
- `symbolic equality domain`: supports new join operation in
quadratic running time

igcontreras commented:
It seems that you are adding new commits by @caballa and others to master, can you update so that we only see the changes of your commits in the review?

igcontreras commented on include/crab/domains/symbolic_variable_eq_domain.hpp:
```diff
@@ -0,0 +1,1031 @@
+  return SYMBVAR(term::term_op_val_generator_t::get_next_val());
+}
+
+template <class Domain> class equivalence_class {
```
If `equivalence_class` is not going to be used outside of this file, you can hide it and make it a struct. Then you don't need to make getters and setters

igcontreras commented on include/crab/domains/symbolic_variable_eq_domain.hpp:
```diff
@@ -0,0 +1,1031 @@
+#define SVEQ_COUNT_STATS(NAME) CRAB_DOMAIN_COUNT_STATS(NAME, 0)
+
+/// @brief An abstract domain represents equalities used in analyses such as
+/// allocation-site abstraction or domain reduction. In short, giving two
```
fix grammar

igcontreras commented on include/crab/domains/symbolic_variable_eq_domain.hpp:
```diff
@@ -0,0 +1,1031 @@
+  //            ┌─┐      ┌─┐          ┌─┐
+  //            │2│      │5│          │3│
+  //            └─┘      └─┘          └─┘
+  // Note that, { 6 } |-> #var3 will be removed after normalization since
```
update this comment, this is not the case anymore, right?

igcontreras commented on include/crab/domains/symbolic_variable_eq_domain.hpp:
```diff
@@ -0,0 +1,1031 @@
+    return res;
+  }
+
+  element_set_t get_all_members() const {
```
not used, remove

igcontreras commented on include/crab/domains/symbolic_variable_eq_domain.hpp:
```diff
@@ -0,0 +1,1031 @@
+  /// @brief union two sets
+  /// @param x one set in domain
+  /// @param y one set in domain
+  void union_sets(const element_t &x, const element_t &y) {
```
This doesn't seem to be used but it looks like `union` operation. Use this instead of overwritting values in `m_parents` directly (see the example in `add`). Try to find all the places where you should be using `union` (`meet` probably)

igcontreras commented on include/crab/domains/symbolic_variable_eq_domain.hpp:
```diff
@@ -0,0 +1,1031 @@
+  /// @param y an element may exists
+  /// @note  if x does not exist, nothing changed.
+  ///        if y \in some cls, forget it from cls, add y into x's class
+  void add(const element_t &x, const element_t &y) {
```
why not call this one assign

igcontreras commented on include/crab/domains/symbolic_variable_eq_domain.hpp:
```diff
@@ -0,0 +1,1031 @@
+    // Return true if *this is a refined partitioning of o
+    //    forall cls_x \in this. exists cls_y \in e ::
+    //      cls_y <= cls_x
+    equivalence_class_elems_t left_equiv_classes =
```
This is not efficient.

The L <= R operation should check that all equalities that exist in R, exist in L as well (but L may have additional equalities). So you do not need to build equivalence classes of L, once you have an equivalence class of R, simply check that these equalities exists in L.

As soon as you find the first equality from R that is not in L, the answer to <= is false, you can return false and do not compute for the rest of the classes.

With this, you can avoid building all equivalence classes for left and right.

igcontreras commented on include/crab/domains/symbolic_variable_eq_domain.hpp:
```diff
@@ -0,0 +1,1031 @@
+  /// @param is_not_rhs if true the assignment is lhs := not(rhs)
+  void assign_bool_var(const element_t &lhs, const element_t &rhs,
+                       bool is_not_rhs) {
+    // WARN: this should not happen
```
Add `SASSERT(false && "this domain does not treat equalities between boolean vars")`

But why shouldn't it happen? Maybe for the object domain it is not expected to happen.

igcontreras commented on include/crab/domains/symbolic_variable_eq_domain.hpp:
```diff
@@ -0,0 +1,1031 @@
+  }
+
+  /// @brief expand what x equals to to y. This is equivalent to add
+  /// @param x the orginal variable in some class
```
typo

igcontreras commented on include/crab/domains/symbolic_variable_eq_domain.hpp:
```diff
@@ -0,0 +1,1031 @@
+  /// @brief Check whether domain constains element v in some class
+  /// @param v an element
+  /// @return true if the elment exists; otherwise, false.
+  bool contains(const element_t &v) const {
```
I would rename this to `has_representative` or `has_definition`

igcontreras commented on include/crab/domains/symbolic_variable_eq_domain.hpp:
```diff
@@ -0,0 +1,1031 @@
+    if (contains(y)) {
+      *this -= y;
+    }
+    element_t rep_x = find(x);
```
This should be done by calling the `union` operation. You are not updating the size of the class

igcontreras commented on include/crab/domains/symbolic_variable_eq_domain.hpp:
```diff
@@ -0,0 +1,1031 @@
+  /// @param absval a domain value
+  /// @note  if x does not exist, create an class with element x and absval
+  ///        if x \in some cls, update current domain value by absval
+  void set(const element_t &x, domain_t absval) {
```
`set_class_value` instead?

igcontreras commented on include/crab/domains/symbolic_variable_eq_domain.hpp:
```diff
@@ -0,0 +1,1031 @@
+  /// @brief get the domain value stored in current equivalent class
+  /// @param v an element in some set
+  /// @return null if !contains(x), otherwise, returns a shared pointer to value
+  std::shared_ptr<domain_t> get(const element_t &x) {
```
`get_class_value`?

igcontreras commented on include/crab/domains/symbolic_variable_eq_domain.hpp:
```diff
@@ -0,0 +1,1031 @@
+  /// @brief get the domain value stored in current equivalent class
+  /// @param v an element in some set
+  /// @return null if !contains(x), otherwise, returns a shared pointer to value
+  std::shared_ptr<const domain_t> get(const element_t &x) const {
```
put this together with the other `get`

igcontreras commented on include/crab/domains/symbolic_variable_eq_domain.hpp:
```diff
@@ -0,0 +1,1031 @@
+  /// keep equalities appeared either in left or in right states.
+  /// Formally, forall cls_x \in left. forall cls_y \in right ::
+  ///           cls_x ^ cls_y != \empty => cls_x union cls_y
+  this_domain_t meet(const this_domain_t &left,
```
Building the classes from left and right is not efficient. Try doing this:

Initialize `res` as a copy of `left` (or `right`). For every equality `a = b` in `right` (or `left`), call `res.union(a,b)`.

What about abstract values of the classes, are you doing the meet for those?

igcontreras commented on include/crab/domains/symbolic_variable_eq_domain.hpp:
```diff
@@ -0,0 +1,1031 @@
+  /// The computation follows the structure of unordered_map.
+  /// The operation requires to construct a new unordered_map.
+  /// TODO: to improve sharable results, we need to change data structure.
+  this_domain_t join(const this_domain_t &left,
```
what about the abstract values of equivalence classes? is join being done?

igcontreras commented on include/crab/domains/symbolic_variable_eq_domain.hpp:
```diff
@@ -0,0 +1,1031 @@
+  /// @param v an element
+  /// @note  if v \in cls \land v is representative, select a new representative
+  ///        before removing v.
+  void operator-=(const element_t &v) {
```
Decrease the size of the class of `v`.

igcontreras commented on include/crab/domains/symbolic_variable_eq_domain.hpp:
```diff
@@ -0,0 +1,1031 @@
+  /// @param elements a vector of elements required to keep
+  /// @note  After project operation, the state only keeps equivalence classes
+  ///        for element in vector elements
+  void project(const std::vector<element_t> &elements) {
```
Is this maintaining the sizes of the classes?

igcontreras commented on include/crab/domains/symbolic_variable_eq_domain.hpp:
```diff
@@ -0,0 +1,1031 @@
+          break;
+        }
+      }
+      element_t new_rep = keep_rep ? rep : s[0];
```
Why create new representatives? Can't you just forget what is not in `elements`?

igcontreras commented on include/crab/domains/symbolic_variable_eq_domain.hpp:
```diff
@@ -0,0 +1,1031 @@
+      CRAB_ERROR(domain_name(),
+                 "::rename with input vectors of different sizes");
+    }
+    for (unsigned i = 0, size = old_elements.size(); i < size; ++i) {
```
use iterators, they are more efficient

igcontreras commented on include/crab/domains/symbolic_variable_eq_domain.hpp:
```diff
@@ -0,0 +1,1031 @@
+          ++it;
+        }
+        if (parent_old_v) {
+          // insert a key-value pair for new_v
```
is it possible to replace a key instead of remove and insert?

igcontreras commented on include/crab/domains/symbolic_variable_eq_domain.hpp:
```diff
@@ -0,0 +1,1031 @@
+    for (auto &kv : equiv_classes) {
+      const element_t &rep = kv.first;
+      const element_set_t &s = kv.second;
+      if (s.size() == 1) {
```
This is not what we talked about. If our equality domains are spread out, the value of the class matters even if the class is of size 1

igcontreras reviewed (commented):
First pass over equality domain

igcontreras commented on include/crab/domains/uf_domain.hpp:
Are you using this domain? Should this be removed?

caballa commented:
Too old. we will need a new one
